### ISSUE TITLE
[clangd] clangd crash when indexing clang/test/SemaCXX/fold_expr_typo.cpp

### ISSUE THREAD

**Issue by yronglin** [clangd/clangd#1692]
clangd crash when indexing clang/test/SemaCXX/fold_expr_typo.cpp

```
// RUN: %clang_cc1 -fsyntax-only -verify -std=c++17 %s

template <typename... T>
void foo(T &&...Params) {
  foo<T>(Unknown); // expected-error {{expression contains unexpanded parameter pack 'T'}}\
                      expected-error {{use of undeclared identifier 'Unknown'}}
  ((foo<T>(Unknown)), ...); // expected-error {{use of undeclared identifier 'Unknown'}}
}

template <typename... U> struct A {
  template <typename... T> void foo(T &&...Params) {
    foo<T>((... + static_cast<U>(1))); // expected-error {{expression contains unexpanded parameter pack 'T'}}
  }
};

```

**Logs**

```console
#0 0x00007f6260d03661 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib/llvm/15/bin/../lib64/libLLVM-15.so+0xf03661)
 #1 0x00007f6260d01064 llvm::sys::RunSignalHandlers() (/usr/lib/llvm/15/bin/../lib64/libLLVM-15.so+0xf01064)
 #2 0x00007f6260d011d6 (/usr/lib/llvm/15/bin/../lib64/libLLVM-15.so+0xf011d6)
 #3 0x00007f625fc64f60 (/lib64/libc.so.6+0x37f60)
 #4 0x00007f62689a8684 clang::TypoCorrectionConsumer::getNextCorrection() (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0x17a8684)
 #5 0x00007f62688c5ce2 (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0x16c5ce2)
 #6 0x00007f6268904094 (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0x1704094)
 #7 0x00007f6268907d8e (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0x1707d8e)
 #8 0x00007f626890d005 (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0x170d005)
 #9 0x00007f626890641f (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0x170641f)
#10 0x00007f62688cf82c (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0x16cf82c)
#11 0x00007f62688d08e1 clang::Sema::CorrectDelayedTyposInExpr(clang::Expr*, clang::VarDecl*, bool, llvm::function_ref<clang::ActionResult<clang::Expr*, true> (clang::Expr*)>) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0x16d08e1)
#12 0x00007f62688d1671 clang::Sema::ActOnFinishFullExpr(clang::Expr*, clang::SourceLocation, bool, bool) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0x16d1671)
#13 0x00007f6268af89ec clang::Sema::ActOnExprStmt(clang::ActionResult<clang::Expr*, true>, bool) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0x18f89ec)
#14 0x00007f6267e5646c clang::Parser::ParseExprStatement(clang::Parser::ParsedStmtContext) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0xc5646c)
#15 0x00007f6267e57552 clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::ParsedAttributes&, clang::ParsedAttributes&) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0xc57552)
#16 0x00007f6267e57cf3 clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0xc57cf3)
#17 0x00007f6267e589a8 clang::Parser::ParseCompoundStatementBody(bool) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0xc589a8)
#18 0x00007f6267e5bd3a clang::Parser::ParseFunctionStatementBody(clang::Decl*, clang::Parser::ParseScope&) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0xc5bd3a)
#19 0x00007f6267e82eaf clang::Parser::ParseFunctionDefinition(clang::ParsingDeclarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::LateParsedAttrList*) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0xc82eaf)
#20 0x00007f6267e6b235 clang::Parser::ParseSingleDeclarationAfterTemplate(clang::DeclaratorContext, clang::Parser::ParsedTemplateInfo const&, clang::ParsingDeclRAIIObject&, clang::SourceLocation&, clang::ParsedAttributes&, clang::AccessSpecifier) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0xc6b235)
#21 0x00007f6267e745e3 clang::Parser::ParseTemplateDeclarationOrSpecialization(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributes&, clang::AccessSpecifier) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0xc745e3)
#22 0x00007f6267e7484a clang::Parser::ParseDeclarationStartingWithTemplate(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributes&, clang::AccessSpecifier) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0xc7484a)
#23 0x00007f6267dcb870 clang::Parser::ParseDeclaration(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributes&, clang::ParsedAttributes&, clang::SourceLocation*) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0xbcb870)
#24 0x00007f6267e84abc clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsingDeclSpec*) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0xc84abc)
#25 0x00007f6267e86277 clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0xc86277)
#26 0x00007f6267e866af clang::Parser::ParseFirstTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0xc866af)
#27 0x00007f6267da5568 clang::ParseAST(clang::Sema&, bool, bool) (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0xba5568)
#28 0x00007f6269872e59 clang::FrontendAction::Execute() (/usr/lib/llvm/15/bin/../lib64/libclang-cpp.so.g1+0x2672e59)
#29 0x0000563f1a29e39f (/usr/lib/llvm/15/bin/clangd+0x92939f)
#30 0x0000563f1a304592 (/usr/lib/llvm/15/bin/clangd+0x98f592)
#31 0x0000563f1a2fb14c (/usr/lib/llvm/15/bin/clangd+0x98614c)
#32 0x0000563f1a2fc518 (/usr/lib/llvm/15/bin/clangd+0x987518)
#33 0x0000563f1a4aee37 (/usr/lib/llvm/15/bin/clangd+0xb39e37)
#34 0x00007f625fcb1fe4 (/lib64/libc.so.6+0x84fe4)
#35 0x00007f625fd3214c (/lib64/libc.so.6+0x10514c)
Signalled during AST worker action: Build AST
  Filename: /home/yrong/Developer/llvm/llvm-project/clang/test/SemaCXX/fold_expr_typo.cpp
  Directory: /home/yrong/Developer/llvm/rel
  Command Line: /usr/x86_64-pc-linux-gnu/gcc-bin/12/c++ --driver-mode=g++ -DGTEST_HAS_RTTI=0 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -I/home/yrong/Developer/llvm/rel/tools/clang/tools/amdgpu-arch -I/home/yrong/Developer/llvm/llvm-project/clang/tools/amdgpu-arch -I/home/yrong/Developer/llvm/llvm-project/clang/include -I/home/yrong/Developer/llvm/rel/tools/clang/include -I/home/yrong/Developer/llvm/rel/include -I/home/yrong/Developer/llvm/llvm-project/llvm/include -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -fno-common -Woverloaded-virtual -fno-strict-aliasing -O3 -DNDEBUG -fno-exceptions -funwind-tables -fno-rtti -c -std=c++17 -resource-dir=/usr/lib/llvm/15/bin/../../../../lib/clang/15.0.7 -- /home/yrong/Developer/llvm/llvm-project/clang/test/SemaCXX/fold_expr_typo.cpp
  Version: 1
[Error - 12:40:24 AM] The Clang Language Server server crashed 5 times in the last 3 minutes. The server will not be restarted. See the output for more information.
```

**System information**

Output of `clangd --version`:
```console
clangd version 15.0.7
Features: linux
Platform: x86_64-pc-linux-gnu
```
Editor/LSP plugin:

Operating system:
Linux

**Comment by zyn0217** [comment]
I think this should be fixed by [D145892](https://reviews.llvm.org/D145892).

**Comment by yronglin** [comment]
> I think this should be fixed by [D145892](https://reviews.llvm.org/D145892). (Dupe of [llvm/llvm-project#61326](https://github.com/llvm/llvm-project/issues/61326))

That works! Awesome,I have build clang from trunk and it's no longer crash!
